### DISCUSSION TITLE
Kalman Filters as discrete DBN

### DISCUSSION THREAD

**Discussion by FelipeGiro** [pgmpy/pgmpy#1451] [Ideas]
Imagine I have a classical [Kalman Filter](https://www.wikiwand.com/en/Kalman_filter#/Details), with prediciton and update. 

```
F, Q = 1.2, 0.0
x0, P0 = 5.5, 1.0
H, z, R = 1.0, 5.8, 0.5

print("Priori state:", x0, P0)

f = lambda x: F*x

x = f(x0)
P = F*P0*F + Q

print("Prediction:", x, P)

print("+ observation:", z, R)

y = z - H*x
S = H*P*H + R
K = P*H/S
x = x + K*y
P =(1 - K*H)*P

print("Update:", x, P)
```

**The question is:** can I use `pgmpy.models.DynamicBayesianNetwork` using discrete CDP to perform Kalman Filter?

I can understand the I can use a `TabularCPD` to represent the transition through time (I call here `transition_matrix`) and the initial state of x using a probability mass function (here `initial_cpd `). Lets say I can compute it easily through Monte Carlo Simulation (`MCS`).

So, the code would be more or less like this:

```
from pgmpy.models import DynamicBayesianNetwork as DBN
from pgmpy.inference import dbn_inference
from pgmpy.factors.discrete.CPD import TabularCPD

import numpy as np
from scipy.stats import norm

dbn = DBN()
dbn.add_edges_from([
    (('x', 0), ('x', 1),
     ('x', 1), ('z', 1))
    ])

disc,= np.linspace(0.0, 10.0, num=100)

initial_pmf = np.diff(norm.cdf(disc, x0, P0**0.5))/delta
transition_matrix = MCS(f)

initial_cpd = TabularCPD(('x', 0), 100, initial_pmf)
transition_cpd = TabularCPD(('x', 1), 100, transition_matrix, 
                            evidence=[('x',0)])
```

Now, the problem how I model the update step with the observation `z`. Does it would be a Tabular CDP dependent on `x`?

```
obs_cdp = TabularCPD(('z', 1), 100, something, 
                     evidence=[('x',1)])
```
If yes, `something `would be 100x100 matrix.

For the filtering task, I would multiply the updated result by the observation PMF and normalize it:

```
updated_pmf = predicted_pmf*obs_pmf
updated_pmf /= updated_pmf.sum()
```

**Why I am interested in that?** Because I want do this inference task with any distribution and any function.

I am wondering if I can use `pgmpy `to model this problem: https://doi.org/10.1061/(ASCE)EM.1943-7889.0000024
The initial state is a discretized exponential distribution, the function is a fracture mechanics equation (non-linear), and the observation is a discretized normal distribution.

So, there are two things I don't see how to do with `pgmpy`:
- build the `transition_matrix `from a function `f`;
- update step with a observation `z` with `R` uncertainty.

Am I missing something? If not, I would be interesting implement it in `pgmpy`?

**Comment by ankurankan** [comment]
I am not very familiar with the details of Kalman filter but I tried to read up a bit on it. It does seem to me that it should be possible to implement it in pgmpy. 

> build the transition_matrix from a function f;

Since the state transitions work as a Markov Chain, I think you can compute the transition matrix from the CPDs. We use a similar function for doing Gibbs sampling: https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/sampling/Sampling.py#L377. 

> update step with a observation z with R uncertainty.

I think this should be equivalent to providing virtual evidence during inference. We do have an option for this for normal Bayesian Network inference (virtual_evidence): https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/inference/ExactInference.py#L226. We will have to probably do something similar for the case of Dynamic Bayesian Network inference as well. Implementing this should be quite simple as it just requires some network modification during inference: https://link.springer.com/chapter/10.1007/978-3-642-31718-7_5
